File: python/131-140/Copy List with Random Pointer.py
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if head==None:
            return None
        origin,rori = {},{}
        new={}
        count=0
        p=head
        while p!=None:
            temp=RandomListNode(p.label)
            origin[count]=p
            rori[p]=count
            new[count]=temp
            count+=1
            p=p.next
        for i in range(count):
            if i ==count-1:
                new[count-1].next=None
            else:
                new[i].next=new[i+1]
            if origin[i].random==None:
                new[i].random=None
            else:
                new[i].random=new[rori[origin[i].random]]
        return new[0]

File: Cpp/130-140/Copy List with Random Pointer.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        if(head==NULL) return NULL;
        RandomListNode* temp=head;
        while(temp!=NULL){
            RandomListNode *t = new RandomListNode(temp->label);
            t->next = temp->next;
            temp->next=t;
            temp = temp->next->next;
        }
        RandomListNode thead(0);
        temp = &thead;
        RandomListNode* cur1;
        for(cur1=head; cur1; cur1=cur1->next->next) {
      cur1->next->random = cur1->random? cur1->random->next:NULL;
    }
    for(cur1=head; cur1; cur1=cur1->next) {
      temp = temp->next = cur1->next;
      cur1->next = temp->next;
    }
         temp->next = NULL;
        return thead.next;
    }
};
